File: Python/_127_Word_Ladder/solution.py
class Solution:
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        queue = [(beginWord, 1)]
        words = set(wordList)
        visited = set()   
        while queue:
            word, dist = queue.pop(0)
            if word == endWord:
                return dist
            for i in range(len(word)):
                for j in 'abcdefghijklmnopqrstuvwxyz':
                    tmp = word[:i] + j + word[i + 1:]
                    if tmp not in visited and tmp in words:
                        queue.append((tmp, dist + 1))
                        visited.add(tmp)
        return 0

File: Java/_127_Word_Ladder/Solution.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Queue<String> visit = new LinkedList<String>();
        Set<String> words = new HashSet<>();
        words.addAll(wordList);
        visit.offer(beginWord);
        HashMap<String, Integer> maps = new HashMap<String, Integer>();
        maps.put(beginWord, 1);
        if (words.contains(beginWord))
            words.remove(beginWord);
        
        while (!visit.isEmpty()) {
            String top = visit.poll();      
            int level = maps.get(top);
            for (int i = 0; i < top.length(); i++) {
                StringBuilder builder = new StringBuilder(top);
                for (char c = 'a'; c <= 'z'; c++) {
                    builder.setCharAt(i, c);
                    String temp = builder.toString();
                    if (temp.equals(top))
                        continue;
                    else if (temp.equals(endWord) && words.contains(temp))
                        return level + 1;
                    else if (words.contains(temp)) {
                        visit.offer(temp);
                        words.remove(temp);
                        maps.put(temp, level + 1);
                    }
                }
            }
        }
        return 0;
    }
}